image_proc/include/rectify.hpp

```diff
@@ -31,47 +31,43 @@
 *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 *  POSSIBILITY OF SUCH DAMAGE.
 *********************************************************************/
-#ifndef IMAGE_PROC_PROCESSOR_H
-#define IMAGE_PROC_PROCESSOR_H
+#include <thread>
+#include <memory>
+#include <vector>
+#include <string>
 
-#include <opencv2/core/core.hpp>
+#include <rclcpp/rclcpp.hpp>
+#include <image_transport/image_transport.h>
 #include <image_geometry/pinhole_camera_model.h>
-#include <sensor_msgs/Image.h>
-
+#include <cv_bridge/cv_bridge.h>
+#include "../src/visibility.h"
 namespace image_proc {
 
-struct ImageSet
+class RectifyNode : public rclcpp::Node
 {
-  std::string color_encoding;
-  cv::Mat mono;
-  cv::Mat rect;
-  cv::Mat color;
-  cv::Mat rect_color;
-};
+  public:
+   IMAGE_PROC_PUBLIC RectifyNode();
+  private:
+  // ROS communication
+    image_transport::CameraSubscriber sub_camera_;
+    int queue_size_;
+    int interpolation;
+    std::string camera_namespace_;
+    std::string image_rect;
+    std::string image_topic;
+    std::mutex connect_mutex_;
+    image_transport::Publisher pub_rect_;
 
-class Processor
-{
-public:
-  Processor()
-    : interpolation_(cv::INTER_LINEAR)
-  {
-  }
-  
-  int interpolation_;
+    // Dynamic reconfigure
 
-  enum {
-    MONO       = 1 << 0,
-    RECT       = 1 << 1,
-    COLOR      = 1 << 2,
-    RECT_COLOR = 1 << 3,
-    ALL = MONO | RECT | COLOR | RECT_COLOR
-  };
-  
-  bool process(const sensor_msgs::ImageConstPtr& raw_image,
-               const image_geometry::PinholeCameraModel& model,
-               ImageSet& output, int flags = ALL) const;
-};
+    // Processing state (note: only safe because we're using single-threaded NodeHandle!)
```

Comment:
these code is just same as ros branch did, so i am not change it in ros2 version.

Reply:
Sorry, to be clear, this was previously using single threaded NodeHandle's in the ROS1/Nodelet paradigm.  Because this is now inheriting from an rclcpp::Node, we can't make the same guarantees about this only executing in a single-threaded context (people could add the Node to a multi-threaded executor). Since that assumption no longer holds, I believe that we need to add some locking here to guarantee safety.